leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -108,23 +117,102 @@ public T buildResponse(final Response coapResponse) {
         final Request coapRequest = coapRequestBuilder.getRequest();
 
         // Add CoAP request callback
-        coapRequest.addMessageObserver(new AsyncRequestObserver<T>(coapRequest, destination, responseCallback,
-                errorCallback) {
-            @Override
-            public T buildResponse(final Response coapResponse) {
-                // Build LwM2m response
-                final LwM2mResponseBuilder<T> lwm2mResponseBuilder = new LwM2mResponseBuilder<T>(coapRequest,
-                        coapResponse, client, model, observationRegistry);
-                request.accept(lwm2mResponseBuilder);
-                return lwm2mResponseBuilder.getResponse();
-            }
-        });
+        coapRequest.addMessageObserver(
+                new AsyncRequestObserver<T>(coapRequest, destination, responseCallback, errorCallback) {
+                    @Override
+                    public T buildResponse(final Response coapResponse) {
+                        // Build LwM2m response
+                        final LwM2mResponseBuilder<T> lwm2mResponseBuilder = new LwM2mResponseBuilder<T>(coapRequest,
+                                coapResponse, client, model, observationRegistry);
+                        request.accept(lwm2mResponseBuilder);
+                        return lwm2mResponseBuilder.getResponse();
+                    }
+                });
+
+        // Store pending request to cancel it on deregistration
+        addPendingRequest(destination.getRegistrationId(), coapRequest);
 
         // Send CoAP request asynchronously
         final Endpoint endpoint = getEndpointForClient(destination);
         endpoint.sendRequest(coapRequest);
     }
 
+    public synchronized void cancelPendingRequests(String registrationId) {
+        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        if (requests != null) {
+            for (Request coapRequest : requests) {
+                coapRequest.cancel();
+                removePendingRequest(registrationId, coapRequest);
```

Comment:
Did you check, if it's OK to remove something from a collection while "iterating" over it?